<a href="https://colab.research.google.com/github/programmermahi/Machine-Learning-Lab/blob/main/221902083_Md_mohibullah_2no_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np

df = pd.read_csv('/content/mushrooms.csv')
display(df.head(10))


cap-shape cap-surface cap-color bruises odor gill-attachment gill-spacing  \
0         x           s         n       t    p               f            c   
1         x           s         y       t    a               f            c   
2         b           s         w       t    l               f            c   
3         x           y         w       t    p               f            c   
4         x           s         g       f    n               f            w   
5         x           y         y       t    a               f            c   
6         b           s         w       t    a               f            c   
7         b           y         w       t    l               f            c   
8         x           y         w       t    p               f            c   
9         b           s         y       t    a               f            c   

  gill-size gill-color stalk-shape  ... stalk-color-above-ring  \
0         n          k           e  ...                      w   
1         b          k           e  ...                      w   
2         b          n           e  ...                      w   
3         n          n           e  ...                      w   
4         b          k           t  ...                      w   
5         b          n           e  ...                      w   
6         b          g           e  ...                      w   
7         b          n           e  ...                      w   
8         n          p           e  ...                      w   
9         b          g           e  ...                      w   

  stalk-color-below-ring veil-type veil-color ring-number ring-type  \
0                      w         p          w           o         p   
1                      w         p          w           o         p   
2                      w         p          w           o         p   
3                      w         p          w           o         p   
4                      w         p          w           o         e   
5                      w         p          w           o         p   
6                      w         p          w           o         p   
7                      w         p          w           o         p   
8                      w         p          w           o         p   
9                      w         p          w           o         p   

  spore-print-color population habitat class  
0                 k          s       u     p  
1                 n          n       g     e  
2                 n          n       m     e  
3                 k          s       u     p  
4                 n          a       g     e  
5                 k          n       g     e  
6                 k          n       m     e  
7                 n          s       m     e  
8                 k          v       g     p  
9                 k          s       m     e  

[10 rows x 23 columns]

In [ ]:
display(df.shape)

(8124, 23)

In [ ]:
display(df.isnull().sum())

cap-shape                   0
cap-surface                 0
cap-color                   0
bruises                     0
odor                        0
gill-attachment             0
gill-spacing                0
gill-size                   0
gill-color                  0
stalk-shape                 0
stalk-root                  0
stalk-surface-above-ring    0
stalk-surface-below-ring    0
stalk-color-above-ring      0
stalk-color-below-ring      0
veil-type                   0
veil-color                  0
ring-number                 0
ring-type                   0
spore-print-color           0
population                  0
habitat                     0
class                       0
dtype: int64

In [ ]:
categorical_cols = df.select_dtypes(include=['object']).columns
df_encoded = pd.get_dummies(df, columns=categorical_cols, drop_first=True)
display(df_encoded.head())

cap-shape_c  cap-shape_f  cap-shape_k  cap-shape_s  cap-shape_x  \
0        False        False        False        False         True   
1        False        False        False        False         True   
2        False        False        False        False        False   
3        False        False        False        False         True   
4        False        False        False        False         True   

   cap-surface_g  cap-surface_s  cap-surface_y  cap-color_c  cap-color_e  ...  \
0          False           True          False        False        False  ...   
1          False           True          False        False        False  ...   
2          False           True          False        False        False  ...   
3          False          False           True        False        False  ...   
4          False           True          False        False        False  ...   

   population_s  population_v  population_y  habitat_g  habitat_l  habitat_m  \
0          True         False         False      False      False      False   
1         False         False         False       True      False      False   
2         False         False         False      False      False       True   
3          True         False         False      False      False      False   
4         False         False         False       True      False      False   

   habitat_p  habitat_u  habitat_w  class_p  
0      False       True      False     True  
1      False      False      False    False  
2      False      False      False    False  
3      False       True      False     True  
4      False      False      False    False  

[5 rows x 96 columns]

In [ ]:
from sklearn.model_selection import train_test_split


X = df_processed.drop(columns=['class_p', 'class_e'])
y = df_processed[['class_p', 'class_e']]


X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

display(f"X_train shape: {X_train.shape}")
display(f"X_test shape: {X_test.shape}")
display(f"y_train shape: {y_train.shape}")
display(f"y_test shape: {y_test.shape}")

'X_train shape: (6499, 116)'

'X_test shape: (1625, 116)'

'y_train shape: (6499, 2)'

'y_test shape: (1625, 2)'

In [ ]:
df=pd.DataFrame()

In [ ]:
from sklearn.naive_bayes import GaussianNB

y_train_single_column = y_train['class_p']

model = GaussianNB()

model.fit(X_train, y_train_single_column)

GaussianNB()

In [ ]:
from sklearn.metrics import accuracy_score, confusion_matrix

y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test['class_p'], y_pred)
print(f"Accuracy: {accuracy}")

conf_matrix = confusion_matrix(y_test['class_p'], y_pred)

print("Confusion Matrix:")
print(conf_matrix)

Accuracy: 0.9601181683899557
Confusion Matrix:
[[959  81]
 [  0 991]]


In [ ]:
y_pred = model.predict(X_test)
df = pd.DataFrame({'Actual': y_test['class_p'], 'Predicted': y_pred})
#display(df.head(20))
df

Actual  Predicted
1971   False      False
6654    True       True
5606    True       True
3332   False      False
6988    True       True
...      ...        ...
7374    True       True
1149   False      False
4999    True       True
7497    True       True
3341    True       True

[1625 rows x 2 columns]

**Model Training and Evaluation**

I have trained a Gaussian Naive Bayes classifier on the preprocessed mushroom dataset to predict whether a mushroom is edible or poisonous.

The data was split into training and testing sets (80/20 split). The model was trained on the training data and evaluated on the testing data.

**Results**

The model achieved an accuracy of 96.01% on the test set. The confusion matrix shows the following:

- True Positives (Correctly predicted poisonous): 991
- True Negatives (Correctly predicted edible): 959
- False Positives (Incorrectly predicted poisonous): 81
- False Negatives (Incorrectly predicted edible): 0

This indicates that the model is very good at identifying edible mushrooms (no false negatives), but it does have some false positives (predicting poisonous when it's edible). This is a good result for this type of problem, as it's safer to misclassify an edible mushroom as poisonous than vice versa.

# Task
Build a Naive Bayes classifier to predict whether a mushroom is edible or poisonous.

## Explain the model training code

### Subtask:
Describe the code in cell `WlQy_CnDZUp6` which trains the Gaussian Naive Bayes model.


**Reasoning**:
Describe the code in cell `WlQy_CnDZUp6` according to the instructions provided.



In [ ]:
# Step 1: Explain the import
print("Cell WlQy_CnDZUp6 imports the GaussianNB class from sklearn.naive_bayes.")

# Step 2: Explain the target variable extraction
print("The code extracts a single column 'class_p' from y_train to use as the target variable for the Naive Bayes model.")

# Step 3: Describe model instantiation
print("A GaussianNB model is instantiated.")

# Step 4: Explain the model training
print("The .fit() method is called on the instantiated model, using X_train and the extracted single column target variable y_train_single_column to train the model.")

Cell WlQy_CnDZUp6 imports the GaussianNB class from sklearn.naive_bayes.
The code extracts a single column 'class_p' from y_train to use as the target variable for the Naive Bayes model.
A GaussianNB model is instantiated.
The .fit() method is called on the instantiated model, using X_train and the extracted single column target variable y_train_single_column to train the model.


## Explain the model evaluation code

### Subtask:
Describe the code in cell `Sx5hCwgfZY9i` which evaluates the model.


**Reasoning**:
Print the explanations for each step of the model evaluation code as requested in the instructions.



In [ ]:
print("Cell Sx5hCwgfZY9i imports the accuracy_score and confusion_matrix functions from sklearn.metrics.")
print("The code uses the trained model's .predict() method to generate predictions (y_pred) on the test data (X_test).")
print("The accuracy_score function is used to calculate the accuracy of the model by comparing the predicted values (y_pred) to the true test labels (y_test['class_p']).")
print("The confusion_matrix function is used to create a confusion matrix comparing the true test labels (y_test['class_p']) and the predicted values (y_pred).")
print("The calculated accuracy and the confusion matrix are printed to the console.")

Cell Sx5hCwgfZY9i imports the accuracy_score and confusion_matrix functions from sklearn.metrics.
The code uses the trained model's .predict() method to generate predictions (y_pred) on the test data (X_test).
The accuracy_score function is used to calculate the accuracy of the model by comparing the predicted values (y_pred) to the true test labels (y_test['class_p']).
The confusion_matrix function is used to create a confusion matrix comparing the true test labels (y_test['class_p']) and the predicted values (y_pred).
The calculated accuracy and the confusion matrix are printed to the console.


## Summary:

### Data Analysis Key Findings

*   The code imported `GaussianNB` from `sklearn.naive_bayes` to train the model.
*   The target variable 'class\_p' was extracted from `y_train` for training.
*   A `GaussianNB` model was instantiated and trained using `X_train` and the extracted target variable.
*   Model evaluation involved importing `accuracy_score` and `confusion_matrix` from `sklearn.metrics`.
*   Predictions (`y_pred`) were generated on `X_test` using the trained model's `.predict()` method.
*   `accuracy_score` was used to calculate the model's accuracy by comparing `y_pred` with `y_test['class_p']`.
*   A confusion matrix was created using `confusion_matrix` to compare `y_test['class_p']` and `y_pred`.
*   The calculated accuracy and the confusion matrix were printed.

### Insights or Next Steps

*   Evaluate the model's performance based on the printed accuracy and confusion matrix to understand its effectiveness in classifying mushrooms.
*   Consider exploring other Naive Bayes variants or different classification algorithms if the performance of the Gaussian Naive Bayes model is not satisfactory.
